### ISSUE TITLE
StreamBridge.send method propagate IllegalStateException if called with binder

### ISSUE THREAD

**Issue by silvio-didonna** [spring-cloud/spring-cloud-stream#2379]
Using `StreamBridge`'s method `send` with binderName parameter leads to IllegalStateException if there are 2 binders with no defaultCandidate and no setted binding defined in properties.

1. properties contains two or more binder configuration with no defaultCandidate set, in this example `binder1` and `binder2`;
2. topic is defined at runtime, no properties connect it to the binder needed;
3. Call StreamBridge's method `public boolean send(String bindingName, @Nullable String binderName, Object data)` with binderName equals to "binder2"
4. raised exception `java.lang.IllegalStateException: A default binder has been requested, but there is more than one binder available for 'org.springframework.messaging.MessageChannel' : binder1,binder2, and no default binder has been set.`

**Version of the framework**
- Spring Boot 2.6.7
- Spring Cloud 2021.0.2
- Spring Cloud Stream 3.2.3

**Expected behavior**
Method send use the binder name argument sent to select the correct binder as [documentation](https://docs.spring.io/spring-cloud-stream/docs/current/reference/html/spring-cloud-stream.html#_using_specific_binder_type_with_streambridge)

**Additional context**
Works as expected with:
- Spring Boot 2.6.4
- Spring Cloud  2021.0.1
- Spring Cloud Stream 3.2.2

Seems like [this](https://github.com/spring-cloud/spring-cloud-stream/commit/4dae3d032469bb2270b9483f83c959e70bd62b92) commit introduce the code that change the behavior

**Comment by waroir20** [comment]
I am also running into a similar situation, but for mine I want to dynamically decide the topic and whether the payload should be JSON or AVRO.
From debugging it seems like the issue is when the send method tries to look up the "targetType"
        `String targetType = this.resolveBinderTargetType(bindingName, MessageChannel.class, (BinderFactory)this.applicationContext.getBean(BinderFactory.class));`
At this point it is unable to find a specific configuration for that topic/binder and fails.

Any ideas if/how to get past this?

**Comment by sobychacko** [comment]
@silvio-didonna @waroir20 Can you please create a small application where we can reproduce the issue?

**Comment by waroir20** [comment]
> @silvio-didonna @waroir20 Can you please create a small application where we can reproduce the issue?

Yeah, I can. I will do tomorrow.

**Comment by waroir20** [comment]
@sobychacko I have created a sample app with all client related info removed. Let me know if there is something more you would like me to add/elaborate on...
**branch**: main
https://github.com/waroir20/kafka-serializer-publisher

**Comment by sobychacko** [comment]
@silvio-didonna You need to provide the binder name on the actual binding as below.
```
spring.cloud.stream.bindings.process-in-0.binder=binder1 (or binder2)
spring.cloud.stream.bindings.first-out-0.binder=binder1 (or binder2)
spring.cloud.stream.bindings.second-out-0.binder=binder1 (or binder2)
```
Now, the fact that the binder name provided on the `send` method call on `StreamBridge` is not honored could be a bug. We will investigate that. But try to see if the above workaround fixes your issue. 

I also want to point out that when calling `send` on `StreamBridge`, you need to provide the actual binding names as the first argument - in your case, `first-out-0` and `second-out-0` respectively.

**Comment by sobychacko** [comment]
@waroir20 See the comment above. However, your use case looks a bit more complex. I looked at the application you provided. There, it seems like you get the binding (same as the topic name) dynamically through the web request, right? If that is the case, you won't be able to attach the binder name as a property on the binding name. We will look into seeing why the binder name is not honored on the `send` call, as I commented above. In the meantime, as a workaround, somehow provide the property `spring.cloud.stream.bindings.<binding-name>.binder` with the value of the binder name right before the `streamBridge.send` call.

**Comment by silvio-didonna** [comment]
@sobychacko thanks for your reply.

Unfortunately I can't set the topic in properties, it is choosed dynamically

**Comment by sobychacko** [comment]
Ok, we will look into the issue then why `StreamBridge` doesn't honor the binder name.

**Comment by sobychacko** [comment]
@silvio-didonna @waroir20 Thanks for the feedback. I sent a [PR](https://github.com/spring-cloud/spring-cloud-stream/pull/2401) addressing the issue. I will update once it is merged and backported to `3.2.x`.
